Web Inspector: REGRESSION(264158@main) Sources tab never shows files when inspecting JS Context

rcaliman-apple opened the pull request:
#### 1ea561941b3734c55e84711becbc9d2292166228
<pre>
Web Inspector: REGRESSION(264158@main) Sources tab never shows files when inspecting JS Context
<a href="https://bugs.webkit.org/show_bug.cgi?id=258136">https://bugs.webkit.org/show_bug.cgi?id=258136</a>
rdar://110119826

Reviewed by Devin Rousso and Patrick Angle.

When inspecting a JSContext, `WI.pageTarget` is null.
A JSContex&apos;s target type is `WI.TargetType.JavaScript`.

At this point, the `target.mainResource` is `null`,
tripping assertions and causing a crash later down the call stack
when attempting to render the tree element for the assumed worker script.

Fix this by explicitly checking for the target type of `WI.TargetType.Worker`
before attempting to continue rendering source tree elements for worker targets.

* Source/WebInspectorUI/UserInterface/Views/SourcesNavigationSidebarPanel.js:
(WI.SourcesNavigationSidebarPanel.prototype._handleResourceGroupingModeChanged):

Canonical link: <a href="https://commits.webkit.org/265209@main">https://commits.webkit.org/265209@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c25c9365ca88b262fe086156a5ecf9cf5021a083

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10256 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10493 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10757 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11902 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/9878 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12488 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10445 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12821 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10414 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11149 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8641 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12288 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8455 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9285 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/16573 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9555 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9436 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12688 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/9895 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/8027 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/9054 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2460 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/13304 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9728 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Views/SourcesNavigationSidebarPanel.js:
```diff
@@ -2383,7 +2383,7 @@ WI.SourcesNavigationSidebarPanel = class SourcesNavigationSidebarPanel extends W
 
         for (let target of WI.targets) {
-            if (target !== WI.pageTarget) {
+            if (target.type === WI.TargetType.Worker || target.type === WI.TargetType.ServiceWorker) {
```
Good catch! Indeed, inspecting standalone service workers is busted now.
The cause again is `target.mainResource` being `null` at this point.

But I'm confused. 🤔 The assertion in [`WI.SourcesNavigationSidebarPanel._addWorkerTargetWithMainResource()`](https://github.com/WebKit/WebKit/blob/d98c1f9162a0f5cc180204ed11aeded85cc96795/Source/WebInspectorUI/UserInterface/Views/SourcesNavigationSidebarPanel.js#L1310) seems to be expecting this target type.

```
console.assert(target.type === WI.TargetType.Worker || target.type === WI.TargetType.ServiceWorker);
```

@dcrousso can you shine some light into what's going on?

rcaliman-apple replied:
Removed check for `target.type === WI.TargetType.ServiceWorker` and that solves the issue for service workers.

Still interested in learning if this check is sufficient to avoid breaking other inspectable targets.
What would we miss if we used this instead?

```diff
for (let target of WI.targets) {
-            if (target !== WI.pageTarget) {
+            if (target !== WI.mainTarget) {
```

dcrousso replied:
it looks like `target.type === WI.TargetType.ServiceWorker` was erroneously added in [209373@main](https://commits.webkit.org/209373@main).  i think we can/should remove it (in both places)

i think probably the most accurate/correct thing to do here would be to `target.type === WI.TargetType.Worker` or `target instances WI.WorkerTarget`, as checking for `!== WI.pageTarget`/`!== WI.mainTarget` could cause issues in the future if more `WI.Target` are introduced

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/41f9fde6e07997fb2072d3d589591bbd45a4a151. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/15004

dcrousso reviewed (approved):
we may wanna apply this same change in `_addScript` too

i'd also drive-by remove the `WI.TargetType.ServiceWorker` check in `_addWorkerTargetWithMainResource`

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c25c9365ca88b262fe086156a5ecf9cf5021a083. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/15004

rcaliman-apple committed:
Web Inspector: REGRESSION(264158@main) Sources tab never shows files when inspecting JS Context
https://bugs.webkit.org/show_bug.cgi?id=258136
rdar://110119826

Reviewed by Devin Rousso and Patrick Angle.

When inspecting a JSContext, `WI.pageTarget` is null.
A JSContex's target type is `WI.TargetType.JavaScript`.

At this point, the `target.mainResource` is `null`,
tripping assertions and causing a crash later down the call stack
when attempting to render the tree element for the assumed worker script.

Fix this by explicitly checking for the target type of `WI.TargetType.Worker`
before attempting to continue rendering source tree elements for worker targets.

* Source/WebInspectorUI/UserInterface/Views/SourcesNavigationSidebarPanel.js:
(WI.SourcesNavigationSidebarPanel.prototype._handleResourceGroupingModeChanged):

Canonical link: https://commits.webkit.org/265209@main
